Add ClientID SQL Detection (#268)
cossacklabs/product-docs: content/acra/security-controls/client-id-sql-detection/_index.md

---
title: ClientID SQL Detection
weight: 12
---

## ClientID Detection from SQL

{{< hint info >}}

> This feature is available in [Acra Enterprise Edition](/acra/enterprise-edition/) only.

Reviewer: ```suggestion
This feature is available in [Acra Enterprise Edition](/acra/enterprise-edition/) only starting from 0.93.
```

or 0.94?

{{< /hint>}}

> 

Reviewer: we should add 1-2 sentence on why this feature is needed and why users will want to use it. I.e. some use cases would be useful.

AcraServer supports the ability to retrieve the ClientID from SQL in the runtime dynamically.

AcraServer uses a default behavior in which the ClientID is retrieved from TLS

> certificates [`--tls_client_id_from_cert` flag]({{< ref "acra/configuring-maintaining/general-configuration/acra-server/#tls" >}}), provided by the client during handshake.

Reviewer: you know, that you can write simply `[--tls_client_id_from_cert flag](/acra/configuring-maintaining/general-configuration/acra-server/#tls)`, right? :)

> AcraServer uses a default behavior in which the ClientID is retrieved from TLS
> certificates [`--tls_client_id_from_cert` flag]({{< ref "acra/configuring-maintaining/general-configuration/acra-server/#tls" >}}), provided by the client during handshake.

Reviewer: ```suggestion
AcraServer encrypts/decrypts data per client (application, user or group of users). To distinguish the clients, AcraServer uses "ClientID" which it retrieves from TLS certificate provided by the client during handshake (use flag [`--tls_client_id_from_cert`](/acra/configuring-maintaining/general-configuration/acra-server/#tls)).
```

To enable SQL ClientID detection, provide `--client_id_sql_varname=<variable>` flag which set SQL variable name to use
as ClientID in session scope. After SQL ClientID detection is ON, AcraServer expect `SET` SQL command to extract
ClientID in runtime.

AcraServer fully support the SET SQL commands syntax
for [PostgreSQL](https://www.postgresql.org/docs/current/sql-set.html)
and [MySQL](https://dev.mysql.com/doc/refman/5.7/en/set-variable.html#set-variable-user-variables).

Example of PostgreSQL SET commands (`--client_id_sql_varname='app.client_id'` ):

```
set app.client_id = 'client_id' || set app.client_id TO 'client_id'
```

Example of MySQL SET command (`--client_id_sql_varname='@app.client_id'`):

```
set @app.client_id TO 'client_id'
```

## ClientID Ignoring

{{< hint info >}}

> This feature is available in [Acra Enterprise Edition](/acra/enterprise-edition/) only.

Reviewer: ```suggestion
This feature is available in [Acra Enterprise Edition](/acra/enterprise-edition/) only starting from 0.93.
```

or 0.94?

{{< /hint>}}

> 

Reviewer: ```suggestion

Sometimes it's useful NOT to add encryption/decryption for certain applications ("clients" in Acra terms). Now you can configure a list of "Ignored ClientIDs": a traffic coming from them won't be parsed, encrypted or decrypted.
```

> AcraServer also supports the ability to ignore and skip encryption/decryption for specified ClientIDs. TLS-extracted, SQL detected ClientIDs are applicable to ignore by AcraServer.

Reviewer: ```suggestion
AcraServer supports the ability to ignore and skip encryption/decryption for specified ClientIDs. If clientID is extracted from TLS certificate, AcraServer can match it with "Ignored ClientIDs" like and do NOT perform encryption/decryption for this client, send its queries "as is".
```
